File: practice/strings/longestCommonPrefix.py
class Solution:
    def longestCommonPrefix(self, strs: list[str]) -> str:
        if len(strs) == 0:
            return ""

        prelen = len(strs[0])

        for i in range(len(strs) - 1):
            prefix = 0

            for j in range(min(prelen, len(strs[i + 1]))):
                if (strs[i][j] == strs[i + 1][j]):
                    prefix += 1
                else:
                    break

            prelen = min(prelen, prefix)

        prefix = strs[0][:prelen]
        return prefix

s = Solution()
print(s.longestCommonPrefix(["flower","flow","flight"]))

File: practice/strings/longestCommonPrefix.c
#include "stdio.h"
#include "string.h"
#include "stdlib.h"

int min(int a, int b) {
    return a < b ? a : b;
}

char* longestCommonPrefix(char* strs[], int strsSize) {
    if (strsSize == 0) {
        return "";
    }

    int prefix_length = sizeof(strs[0]);

    for (size_t i = 0; i < strsSize - 1; i++)
    {
        int current_prefix = 0;
        for (size_t k = 0; k < prefix_length; k++)
        {
            if (strs[i][k] == strs[i + 1][k]) current_prefix++;
            else break;
        }
        prefix_length = min(current_prefix, prefix_length);
    }

    char* prefix = malloc(prefix_length + 1);
    strncpy(prefix, strs[0], prefix_length);
    prefix[prefix_length] = '\0';
    return prefix;
}

int main() {
    const char* strs[] = {"gleo", "ghna"};
    char* a = longestCommonPrefix(strs, 2);
    printf("%s\n", a);
    return 0;
}
